File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleAvatar.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.unit.dp
import com.smileid.designsystem.SmileDimens
import com.smileid.designsystem.smileProfileHues
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** Initials in a rounded square, or a placeholder without them. The size scales with the font scale, because wrapping the initials renders an ellipse at 2x. */
@Composable
fun UseSmileIDSampleAvatar(
    initials: String,
    modifier: Modifier = Modifier,
    size: Dp = SmileDimens.space40,
    containerColor: Color = smileProfileHues.first(),
) {
    val hasInitials = initials.isNotBlank()
    val diameter = size * LocalDensity.current.fontScale
    Surface(
        modifier = modifier.size(diameter),
        shape = RoundedCornerShape(AVATAR_RADIUS),
        color = if (hasInitials) containerColor else UseSmileIDSampleTheme.colors.avatar.placeholderBackground,
    ) {
        Box(contentAlignment = Alignment.Center) {
            Text(
                text = if (hasInitials) initials else "+",
                style = UseSmileIDSampleTheme.type.avatarFont,
                color = if (hasInitials) {
                    UseSmileIDSampleTheme.colors.avatar.text
                } else {
                    UseSmileIDSampleTheme.colors.avatar.placeholderIcon
                },
                textAlign = TextAlign.Center,
            )
        }
    }
}

/** The avatar fill for a profile at [profileIndex], cycled. Position, not a hash of the initials, which reproduces no design order. */
fun avatarColorForProfile(profileIndex: Int): Color =
    smileProfileHues[profileIndex.coerceAtLeast(0) % smileProfileHues.size]

private val AVATAR_RADIUS = 12.dp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_avatar.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_product_hues.dart';
import '../tokens/smile_tokens.dart';

/// Initials in a rounded square, or a placeholder without them.
class UseSmileIDSampleAvatar extends StatelessWidget {
  /// [size] is the component's own default; a parent whose spec names a different metric passes it.
  const UseSmileIDSampleAvatar({
    required this.initials,
    this.size = SmileDimens.space40,
    this.containerColor,
    super.key,
  });

  /// The initials to draw; blank draws the placeholder instead.
  final String initials;

  /// The square's side before text scaling.
  final double size;

  /// The per-profile fill; the first hue when the caller has no profile to colour by.
  final Color? containerColor;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    final bool hasInitials = initials.trim().isNotEmpty;
    // The square grows with text, because wrapping the initials renders an ellipse at 2x.
    final double side = MediaQuery.textScalerOf(context).scale(size);
    return Container(
      width: side,
      height: side,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        color: hasInitials
            ? (containerColor ?? smileProfileHues.first)
            : colors.avatar.placeholderBackground,
        borderRadius: UseSmileIDSampleShapes.avatar,
      ),
      child: Text(
        hasInitials ? initials : '?',
        textAlign: TextAlign.center,
        style: UseSmileIDSampleType.avatarFont.copyWith(
          color: hasInitials
              ? colors.avatar.text
              : colors.avatar.placeholderIcon,
        ),
      ),
    );
  }
}

/// The avatar fill for a profile at [profileIndex], cycled by list position.
Color avatarColorForProfile(int profileIndex) =>
    smileProfileHues[(profileIndex < 0 ? 0 : profileIndex) %
        smileProfileHues.length];
